server/worker/tasks.py

```diff
@@ -0,0 +1,112 @@
+import uuid
+import traceback
+import json
+import logging
+from datetime import datetime
+from typing import Optional
+from sqlalchemy import update
+
+from ..database import db_session
+from ..models import *  # pylint: disable=wildcard-import
+from ..api import rounds
+from ..util.isoformat import isoformat
+from ..util.jsonschema import JSONDict
+from ..config import FLASK_ENV
+
+
+logger = logging.getLogger("arlo.worker")
+
+
+class UserError(Exception):
+    pass
+
+
+def create_background_task(task_name: TaskName, payload: JSONDict) -> BackgroundTask:
+    task = BackgroundTask(id=str(uuid.uuid4()), task_name=task_name, payload=payload)
+    db_session.add(task)
+    # In testing, we run the task immediately, since we'll only be doing small
+    # tasks and we want to make sure they run in a thread-safe way (i.e. we
+    # don't want tests to interfere with each other's background tasks when
+    # running concurrently).
+    if FLASK_ENV == "test":
+        db_session.flush()
+        run_task(task)
+    return task
+
+
+def serialize_background_task(task: Optional[BackgroundTask]) -> Optional[JSONDict]:
+    if task is None:
+        return None
+
+    if task.error:
+        status = ProcessingStatus.ERRORED
+    elif task.completed_at:
+        status = ProcessingStatus.PROCESSED
+    elif task.started_at:
+        status = ProcessingStatus.PROCESSING
+    else:
+        status = ProcessingStatus.READY_TO_PROCESS
+
+    return {
+        "status": status,
+        "startedAt": isoformat(task.started_at),
+        "completedAt": isoformat(task.completed_at),
+        "error": task.error,
+    }
+
+
+def run_task(task: BackgroundTask) -> bool:
+    print("Run", task.__dict__)
+    if task.started_at:
+        return False
+
+    # Claim this task by updating the `started_at` timestamp in such a way that
+    # it must not have been set before.
+    result = db_session.execute(
+        update(BackgroundTask.__table__)  # pylint: disable=no-member
+        .where(BackgroundTask.id == task.id)
+        .where(BackgroundTask.started_at.is_(None))
+        .values(started_at=datetime.now(timezone.utc))
+    )
+    if result.rowcount == 0:
+        return False
+
+    # If we got this far, task is ours to process.
+    try:
+        task_metadata = json.dumps(
+            dict(id=task.id, task_name=task.task_name, payload=task.payload)
+        )
+        logger.info(f"TASK_START {task_metadata}")
+
+        with db_session.begin_nested():  # pylint: disable=no-member
+            if task.task_name == TaskName.DRAW_SAMPLE:
+                round = Round.query.filter_by(
+                    id=dict(task.payload)["round_id"],
+                    election_id=dict(task.payload)["election_id"],
+                ).one()
+                rounds.draw_sample(round)
```

Comment:
should this be more of a registered handler that we dispatch to, rather than a set of `if/then` statements?

Reply:
definitely would make for a slightly cleaner API. i'll try it out